test/run.d

```diff
@@ -0,0 +1,308 @@
+    ./run.d fail_compilation                                     # runs all tests in fail_compilation
+    ./run.d all                                                  # runs all tests
+    ./run.d clean                                                # remove all test results
+
+Options:
+`, res.options);
+        "\nSee the README.md for a more in-depth explanation of the test-runner.".writeln;
+        return;
+    }
+
+    // parse arguments
+    args.popFront;
+    args2Environment(args);
+
+    // allow overwrites from the environment
+    resultsDir = environment.get("RESULTS_DIR", resultsDir);
+
+    // bootstrap all needed environment variables
+    auto env = getEnvironment;
+
+    // default target
+    if (!args.length)
+        args = ["all"];
+
+    // preprocess
+    auto targets = args.predefinedTargets;
+    // normalize
+    targets = targets.map!(f => format!"%s/%s"(f.absolutePath.dirName.baseName, f.baseName)).array;
+    // filter
+    targets = targets.filterTargets;
+
+    if (targets.length > 0)
+    {
+        auto taskPool = new TaskPool(jobs);
+        scope(exit) taskPool.finish();
+        ensureToolsExists;
+        foreach (target; taskPool.parallel(targets, 1))
+        {
+            auto makeArguments = [scriptDir.buildPath("test_results", "d_do_test"), target];
+            spawnProcess(makeArguments, env, Config.none, scriptDir).wait;
+        }
+    }
+}
+
+/**
+Builds the binary of the tools required by the testsuite
+Does nothing if the tools already exist and are newer than their source
+*/
+void ensureToolsExists()
+{
+    static toolsDir = scriptDir.buildPath("tools");
+    resultsDir.mkdirRecurse;
+    auto tools = [
+        "d_do_test",
+        "sanitize_json",
+    ];
+    // TODO: run in parallel
+    foreach (tool; tools)
+    {
+        auto targetBin = resultsDir.buildPath(tool);
+        auto sourceFile = toolsDir.buildPath(tool ~ ".d");
+        if (targetBin.timeLastModified.ifThrown(SysTime.init) > sourceFile.timeLastModified)
+            writefln("%s is already up-to-date", tool);
+        else
+        {
+            auto command = ["dmd", "-of"~targetBin, sourceFile];
+            writefln("Executing: %-(%s %)", command);
+            spawnProcess(command).wait;
+        }
+    }
+
+    // ensure output directories exist
+    foreach (dir; ["compilable", "fail_compilation", "runnable"])
+        resultsDir.buildPath(dir).mkdirRecurse;
+}
+
+/**
+Goes through the target list and replaces short-hand targets with their expanded version.
+Special targets:
+- clean -> removes resultsDir + immediately stops the runner
+*/
+auto predefinedTargets(string[] targets)
+{
+    static findFiles(string dir)
+    {
+        return scriptDir.buildPath(dir).dirEntries("*{.d,.sh}", SpanMode.shallow).map!(e => e.name);
+    }
+
+    Appender!(string[]) newTargets;
+    foreach (t; targets)
+    {
+        switch (t)
+        {
+            case "clean":
```

Comment:
I had a problem with this, my tab-autocomplete added a trailing slash to the test directory (i.e. `compilable/`).  We could handle this by removing the trailing slash, maybe this?
```D
if (t.endsWith(dirSeparator))
    t = t[0 .. $ - 1];
```